api/auth/azure/azure.go

```diff
@@ -0,0 +1,240 @@
+	VMName            string `json:"name"`
+	VMScaleSetName    string `json:"vmScaleSetName"`
+	SubscriptionID    string `json:"subscriptionId"`
+	ResourceGroupName string `json:"resourceGroupName"`
+}
+
+const (
+	defaultMountPath     = "azure"
+	defaultResourceURL   = "https://management.azure.com/"
+	metadataEndpoint     = "http://169.254.169.254"
+	metadataAPIVersion   = "2021-05-01"
+	apiVersionQueryParam = "api-version"
+	resourceQueryParam   = "resource"
+	clientTimeout        = 10 * time.Second
+)
+
+// NewAzureAuth initializes a new Azure auth method interface to be
+// passed as a parameter to the client.Auth().Login method.
+//
+// Supported options: WithMountPath, WithCloud
+func NewAzureAuth(roleName string, opts ...LoginOption) (*AzureAuth, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("no role name provided for login")
+	}
+
+	a := &AzureAuth{
+		roleName:  roleName,
+		mountPath: defaultMountPath,
+		resource:  defaultResourceURL,
+	}
+
+	// Loop through each option
+	for _, opt := range opts {
+		// Call the option giving the instantiated
+		// *AzureAuth as the argument
+		err := opt(a)
+		if err != nil {
+			return nil, fmt.Errorf("error with login option: %w", err)
+		}
+	}
+
+	// return the modified auth struct instance
+	return a, nil
+}
+
+// Login sets up the required request body for the Azure auth method's /login
+// endpoint, and performs a write to it.
+func (a *AzureAuth) Login(ctx context.Context, client *api.Client) (*api.Secret, error) {
+	jwtResp, err := a.getJWT()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get access token: %w", err)
+	}
+
+	metadataRespJSON, err := getMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get instance metadata: %w", err)
+	}
+
+	loginData := map[string]interface{}{
+		"role":                a.roleName,
+		"jwt":                 jwtResp,
+		"vm_name":             metadataRespJSON.Compute.VMName,
+		"vmss_name":           metadataRespJSON.Compute.VMScaleSetName,
+		"subscription_id":     metadataRespJSON.Compute.SubscriptionID,
+		"resource_group_name": metadataRespJSON.Compute.ResourceGroupName,
+	}
+
+	path := fmt.Sprintf("auth/%s/login", a.mountPath)
+	resp, err := client.Logical().Write(path, loginData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to log in with Azure auth: %w", err)
+	}
+
+	return resp, nil
+}
+
+func WithMountPath(mountPath string) LoginOption {
+	return func(a *AzureAuth) error {
+		a.mountPath = mountPath
+		return nil
+	}
+}
+
+// WithCloud allows you to specify a different cloud environment than
+// the default of Azure Public Cloud. This should match the cloud
+// environment that an administrator configured your Vault server to use.
+//
+// See https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go
+// for a list of valid environments.
+func WithCloud(cloud azure.Environment) LoginOption {
+	return func(a *AzureAuth) error {
+		a.resource = cloud.ResourceManagerEndpoint
```

Comment:
cc @elsesiy I've changed this back to WithResource(url string) because as Austin explained to me, this value is just used to populate the `aud` claim so it needs to match whatever URL the auth backend was configured with. (Which may not actually be an Azure URL, it just depends on what the one who configured Vault set it as.) No Azure API access is specifically given.

Users are still welcome to use the autorest constants you mentioned to populate the string though if they know theirs is an Azure one.

`resource` is admittedly a confusing field all around. I'm keeping the default as the Azure public cloud ARM endpoint just to help those users who probably configured their Vault servers with it since that's what "resource" usually means to them.